wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,370 @@
+          System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(this.waitOnCachedInfoDurationMs);
+
+      while (!isInfoInCache && System.nanoTime() < waitForEndpointInfoTimeoutNano) {
+        cachedInfo =
+            this.pluginService.getInfo(this.customEndpointHostSpec.getHost(), CustomEndpointInfo.class, true);
+        isInfoInCache = cachedInfo != null;
+      }
+
+      if (isInfoInCache) {
+        LOGGER.fine(
+            Messages.get(
+                "CustomEndpointPlugin.foundInfoInCache",
+                new Object[]{ this.customEndpointHostSpec.getHost(), cachedInfo }));
+      }
+    }
+
+    return isInfoInCache;
+  }
+
+  /**
+   * Executes the given method via a pipeline of plugins. If a custom endpoint is being used, a monitor for that custom
+   * endpoint will be created if it does not already exist.
+   *
+   * @param resultClass    The class of the object returned by the {@code jdbcMethodFunc}.
+   * @param exceptionClass The desired exception class for any exceptions that occur while executing the
+   *                       {@code jdbcMethodFunc}.
+   * @param methodInvokeOn The object that the {@code jdbcMethodFunc} is being invoked on.
+   * @param methodName     The name of the method being invoked.
+   * @param jdbcMethodFunc The execute pipeline to call to invoke the method.
+   * @param jdbcMethodArgs The arguments to the method being invoked.
+   * @param <T>            The type of the result returned by the method.
+   * @param <E>            The desired type for any exceptions that occur while executing the {@code jdbcMethodFunc}.
+   * @return The result of the method invocation.
+   * @throws E If an exception occurs, either directly in this method, or while executing the {@code jdbcMethodFunc}.
+   */
+  @Override
+  public <T, E extends Exception> T execute(
+      final Class<T> resultClass,
+      final Class<E> exceptionClass,
+      final Object methodInvokeOn,
+      final String methodName,
+      final JdbcCallable<T, E> jdbcMethodFunc,
+      final Object[] jdbcMethodArgs)
+      throws E {
+    if (this.customEndpointHostSpec == null) {
+      return jdbcMethodFunc.call();
+    }
+
+    createMonitorIfAbsent(this.props);
+
+    // If needed, wait a short time for custom endpoint info to be discovered.
+    boolean isInfoInCache = waitForCustomEndpointInfo();
+
+    if (!isInfoInCache) {
+      SQLException cacheTimeoutException = new SQLException(
+          Messages.get("CustomEndpointPlugin.cacheTimeout",
+              new Object[]{this.waitOnCachedInfoDurationMs, this.customEndpointHostSpec.getHost()}));
+      throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, cacheTimeoutException);
+    }
+
+    return jdbcMethodFunc.call();
+  }
+
+  /**
+   * Closes all active custom endpoint monitors.
+   */
+  public static void closeMonitors() {
+    LOGGER.info(Messages.get("CustomEndpointPlugin.closeMonitors"));
+
+    for (CustomEndpointMonitor monitor : monitors.getEntries().values()) {
```

Comment:
It's better to move it to `monitors.clear()`

Reply:
Yeah looks like monitor.clear already closes the monitors, I'll remove this for loop